### ISSUE TITLE
[Architecture Proposal] Phase 1 Foundation: Dynamic Data-Driven Layer Registry (Single Source of Truth)

### ISSUE THREAD

**Issue by NamanmeetSingh** [c2siorg/tensormap#177]
## Overview

As outlined in the GSoC 2026 Phase 1 Expected Results, TensorMap needs to transition from a hardcoded layer system (requiring 6+ file edits) to a data-driven layer registry. 

To achieve this and easily scale to 15+ layer types (CNNs, RNNs, Regularization), I propose a **Single Source of Truth (SSOT)** architecture. By defining layers in a central JSON registry served by the backend, we can decouple the frontend UI from the backend generation logic.

## Current Bottleneck (The 6-File Problem)

Currently, adding a layer like `Dropout` requires:

1. Hardcoding the button in `Sidebar.jsx`
2. Creating a dedicated `DropoutNode.jsx` UI component
3. Registering the node in `Canvas.jsx`
4. Hardcoding Pydantic schemas in `deep_learning.py`
5. Updating the code generation engine
6. Updating the Keras compilation mapping

## Proposed Architecture: The Unified Registry

I propose creating a `layer_registry.json` on the backend and serving it via a new endpoint (`GET /api/v1/layers`). Both the FastAPI backend and React frontend will use this single contract.

### 1. The JSON Contract (Example)

```json
{
  "layers": {
    "dense": {
      "display_name": "Dense",
      "category": "core",
      "keras_mapping": "tf.keras.layers.Dense",
      "params": {
        "units": { "type": "number", "default": 32, "min": 1, "required": true },
        "activation": { "type": "select", "options": ["relu", "sigmoid", "softmax", "linear"], "default": "relu" }
      }
    },
    "dropout": {
      "display_name": "Dropout",
      "category": "regularization",
      "keras_mapping": "tf.keras.layers.Dropout",
      "params": {
        "rate": { "type": "number", "default": 0.5, "min": 0.0, "max": 1.0, "required": true }
      }
    }
  }
}
````

### 2. Frontend Implementation (Dynamic Rendering)

**Dynamic Sidebar:**
`Sidebar.jsx` fetches the registry on load and maps through layers to generate the draggable buttons, grouping them by category.

**Universal Node Component:**
Instead of a separate `.jsx` file for every layer, we build a single `GenericLayerNode.jsx`. When dropped on the canvas, it looks up its type in the registry and dynamically maps the `params` object to UI inputs (rendering a `<Select>` for `activation` and a number `<Input>` for `units`).

### 3. Backend Implementation (Dynamic Validation)

**Pydantic Dynamic Models:**
Instead of hardcoded models, we use `create_model` in Pydantic to dynamically generate validation schemas based on the JSON registry at server startup.

**Streamlined Code Gen:**
The generation engine loops through the ReactFlow payload, looks up the `keras_mapping` string, and dynamically unpacks the frontend payload into kwargs:

```python
f"{layer.keras_mapping}(**{layer.params})"
```
## Action Plan & Phased Rollout

If the maintainers are aligned with this architecture, I am ready to lead Phase 1 with the following PRs:

* **PR 1:** Define the `layer_registry.json` and build the `GET /api/v1/layers` endpoint.
* **PR 2:** Build the frontend `GenericLayerNode.jsx` to dynamically consume the API and replace the existing hardcoded Dense/Conv2D nodes.
* **PR 3:** Update backend code generation to use the registry dictionary.

@ivantha I'd love your thoughts on this SSOT approach. Does this align with your vision for the Phase 1 Foundation?

**Comment by ivantha** [comment]
Closing: this is already fixed on `main`.

The proposed data-driven layer registry SSOT shipped in full during the GSoC term as a central LayerSpec registry, a layers API, a dynamic sidebar and a GenericLayerNode, with the registry living in a Python module rather than layer_registry.json.

Evidence: tensormap-backend/app/layers/registry.py:108 (LAYER_REGISTRY) and :511 (LAYER_CATEGORIES); tensormap-backend/app/routers/layers.py:13,26,50 (GET /layers/, POST /layers/validate-graph, GET /layers/{type_key}); tensormap-frontend/src/hooks/useLayerRegistry.ts; tensormap-frontend/src/components/nodes/GenericLayerNode.tsx; tensormap-frontend/src/components/DragAndDropCanvas/Sidebar.tsx:158 (commit 8440671, extended by fe448c4).

Thanks for the report. If you still see this on `main`, comment here and we will reopen.
